google/cloud/pubsub/internal/subscriber_connection_impl.cc

```diff
@@ -53,31 +55,54 @@ StatusOr<pubsub::PullResponse> SubscriberConnectionImpl::Pull() {
-        return stub->Pull(context, request);
-      },
-      request, __func__);
-  if (!response) return std::move(response).status();
-  if (response->received_messages_size() != 1) {
-    return internal::InternalError("invalid response, mismatched ID count",
-                                   GCP_ERROR_INFO());
+  auto retry_policy = current.get<pubsub::RetryPolicyOption>()->clone();
+  auto backoff_policy = current.get<pubsub::BackoffPolicyOption>()->clone();
+  google::cloud::Status last_status;
+  google::cloud::StatusOr<google::pubsub::v1::PullResponse> response;
+  while (!retry_policy->IsExhausted()) {
+    grpc::ClientContext context;
+    google::cloud::internal::ConfigureContext(context, current);
+    response = stub_->Pull(context, request);
+
+    if (response && !response->received_messages().empty()) {
+      if (response->received_messages_size() > 1) {
+        return internal::InternalError("invalid response, mismatched ID count",
+                                       GCP_ERROR_INFO());
+      }
+
+      auto received_message =
+          std::move(response->mutable_received_messages()->at(0));
+      auto impl = std::make_unique<pubsub_internal::DefaultPullAckHandler>(
+          background_->cq(), stub_, current, std::move(subscription),
+          std::move(*received_message.mutable_ack_id()),
+          received_message.delivery_attempt());
+      auto message = pubsub_internal::FromProto(
+          std::move(*received_message.mutable_message()));
+      return pubsub::PullResponse{pubsub::PullAckHandler(std::move(impl)),
+                                  std::move(message)};
+    }
+    last_status = std::move(response).status();
+    if (!retry_policy->OnFailure(last_status)) {
+      break;
+    };
+    std::this_thread::sleep_for(backoff_policy->OnCompletion());
+  }
+  if (!last_status.ok()) {
+    return last_status;
+  }
+  if (last_status.ok()) {
+    return google::cloud::internal::UnavailableError("no messages returned",
+                                                     GCP_ERROR_INFO());
+  }
+  if (!retry_policy->IsExhausted()) {
+    return google::cloud::internal::RetryLoopError("Permanent error in",
+                                                   __func__, last_status);
   }
-  auto received_message =
-      std::move(response->mutable_received_messages()->at(0));
-  auto impl = std::make_unique<pubsub_internal::DefaultPullAckHandler>(
-      background_->cq(), stub_, current, std::move(subscription),
-      std::move(*received_message.mutable_ack_id()),
-      received_message.delivery_attempt());
-  auto message = pubsub_internal::FromProto(
-      std::move(*received_message.mutable_message()));
-  return pubsub::PullResponse{pubsub::PullAckHandler(std::move(impl)),
-                              std::move(message)};
+  return google::cloud::internal::RetryLoopError("Retry policy exhausted in",
+                                                 __func__, last_status);
```

Comment:
It's a little hard to tell in the diff, but this looks strange on the face of things.
```
  if (!last_status.ok()) return A;
  if (last_status.ok()) return B;
  if (!retry_policy->IsExhausted()) return C;
  return D;
```
Isn't everything after the second statement unreachable?

Reply:
That's correct. Removed the first if.

Now, it does the following 1) if the last status was ok, it means the rpc returned 0 messages and will give the unavailable error, 2) if the policy is not exhausted, it was a permanent error, 3) return the policy was exhausted.

I tried removing the braces {}, but the formatting is still a little weird. Any suggestions?

Reply:
We generally don't remove braces when the `if` covers multiple lines, so I'd only suggest restoring them.

And one more optional nit ... I'd suggest inverting the final `IsExhausted()` check so that your flow is _"2) if the policy is exhausted, return the policy is exhausted, 3) return permanent error"_, just so the exhausted conditional is more directly related to the exhausted error (and eliminating negations is usually a good thing).

Reply:
Re added the brackets and agreed, swapped the order of the ifs/return